stan/math/rev/mat/functor/integrate_1d_tsc.hpp

```diff
@@ -0,0 +1,179 @@
+
+    /**
+     * Return the numeric integral of a function f given its gradient g.
+     *
+     * @tparam T Type of f.
+     * @tparam G Type of g.
+     * @param f a functor with signature
+     * double (double, std::vector<T_param>) or with signature
+     * double (double, T_param) where the first argument is one being
+     * integrated and the second one is either an extra scalar or vector
+     * being passed to f.
+     * @param g a functor with signature
+     * double (double, std::vector<T_param>, int, std::ostream&) or with
+     * signature double (double, T_param, int, std::ostream&) where the
+     * first argument is onebeing integrated and the second one is
+     * either an extra scalar or vector being passed to f and the
+     * third one selects which component of the gradient vector
+     * is to be returned.
+     * @param a lower limit of integration, must be double type.
+     * @param b upper limit of integration, must be double type.
+     * @param tre target relative error.
+     * @param tae target absolute error.
+     * @param param aditional parameters to be passed to f.
+     * @param msgs stream.
+     * @return numeric integral of function f.
+     */
+    template <typename F, typename G, typename T_param>
+    inline typename scalar_type<T_param>::type
+    integrate_1d_tsc_tscg(const F& f,  const G& g,  const double a,
+                          const double b, const T_param& param,
+                          std::ostream& msgs, const double tre = 1e-6,
+                          const double tae = 1e-6) {
+      check_finite("integrate_1d_tsc", "lower limit", a);
+      check_finite("integrate_1d_tsc", "upper limit", b);
+
+      using std::placeholders::_1;
+
+      // hard case, we want a normalizing factor
+      if (!is_constant_struct<T_param>::value) {
+        size_t N = length(param);
+        std::vector<double> grad(N);
+
+        auto value_of_param = value_of(param);
+
+        for (size_t n = 0; n < N; n++)
+          grad[n] = de_integrator(std::bind<double>(g, _1,
+                                                    value_of_param,
+                                                    static_cast<int>
+                                                    (n+1),
+                                                    std::ref(msgs)),
+                                  a, b, tre, tae);
+
+        double val_ = de_integrator(std::bind<double>(f, _1,
+                                                      value_of_param,
+                                                      std::ref(msgs)),
+                                    a, b, tre, tae);
+
+        operands_and_partials<T_param> ops_partials(param);
+        for (size_t n = 0; n < N; n++)
+          ops_partials.edge1_.partials_[n] += grad[n];
+
+        return ops_partials.build(val_);
+      // easy case, here we are calculating a normalizing constant,
+      // not a normalizing factor, so g doesn't matter at all
+      } else {
```

Comment:
You want to reverse the conditions, return, then not indent.  So rather than
```
if (!special_case) {
  A
} else {
  B
}
```
use
```
if (special_case)
  B;
A;
```
It removes the level of nesting and the level of negation.  Both make the overall structure of the main algorithm clearer.